Data/Profunctor/Product/Class.hs

```diff
@@ -94,7 +95,18 @@ class Profunctor p => ProductProfunctor p where
   f ***! g = (,) `Profunctor.rmap` Profunctor.lmap fst f
                   **** Profunctor.lmap snd g
 
+-- | In the future 'VoidProfunctor' will be a superclass of
```

Comment:
Interesting. I didn't expect that. If we really want to do the superclassing in that direction then we can do it now with no backwards compatibility changes!

Reply:
If we have a superclass relationship between `SumProfunctor` and `VoidProfunctor`, it should be this one. "Providing a unit" is almost always done with a subclass (`Semigroup => Monoid` etc.) as it lets you write laws in terms of superclass operations that you know you have.

Reply:
Yes, that rationale makes sense to me, thanks!  I guess I can also add a superclass of `ProductProfunctor` with method `pureP`. Perhaps I can add `purePP = pureP` as a default and eventually remove `purePP`.

On the other hand this suggests the names need work. The names `UnitProfunctor` and `VoidProfunctor` no longer capture everything the classes do.

`ApplicativeProfunctor` from your "Option 2" now sounds good to me, but I prefer `DecidableProfunctor` to `ConcludeProfunctor`. What do you think?

What do you think?

Reply:
If you have `pureP`, you have `unitP`, so it needs to be on the `UnitProfunctor` subclass, becuase
```haskell
pureP a = dimap (const ()) (const a) unitP
unitP = pureP ()
```

I think I like the `{Sum,Void,Product,Unit}Profunctor` names better, because to me that's the fundamental idea being captured: you're tying pairs of profunctors together in interesting ways, and sometimes you have units for these operations. _From there_, you can recover a bunch of familiar-looking operations from the `Applicative`/`Divisible`/`Decidable` world, but I don't feel like any of them are compelling enough to take over the class names. Also, it makes the package name confusing if there's no `ProductProfunctor` class any more.